### ISSUE TITLE
Performance issue in module tasks due to contagious delays

### ISSUE THREAD

**Issue by ljwo** [taskflow/taskflow#649]
**Describe the bug**
This issue is based on the same root cause and the same mechanism as reported in https://github.com/taskflow/taskflow/issues/647, but since in the present scenario there are no semaphores and no deadlocks, just performance issues, I am opening a separate ticket.  

**To Reproduce**
Code illustrating the issue is available at https://github.com/ljwo/taskflow-semaphores with the new scenario in namespace `time1`.

When the thread that started a module task steals another that takes a long time, the original line of execution cannot continue and must wait until that stolen unrelated long task is finished.
![time1](https://github.com/user-attachments/assets/69c187e8-fa04-4416-8e15-1ae80ce07b8a)
In this case, which is similar to case 2 of https://github.com/taskflow/taskflow/issues/647, the thread that started the module task likely goes stealing after having executed `mod_f2` and likely steals `f5_start_timed`. In the meantime, the completion of the `m1` module waits for `mod_f1` to finish, and this is set up to take about 40ms total, so 20ms from the point the thread went stealing. If `f5_work` is set up to take a short time, say 5ms, no problematic delays are observed:
![5ms](https://github.com/user-attachments/assets/d8b7781f-ff97-4e79-8bb5-a38ecda1bff7)
so `f1` takes roughly the expected 40ms, but when `f5_work` is set up to take a time longer than that remaining in the `f1` line, like 500ms, then this is what happens:
![500ms](https://github.com/user-attachments/assets/05fd5eac-2e2d-4a59-a9e2-99db694b1716)
`f1` now takes half a second to complete because the thread happened to execute some completely unrelated long lasting task. This is unfortunate when performing calculations, but completely disastrous if you try to use taskflow for process control.

**Comment by tsung-wei-huang** [comment]
@ljwo thank you for the issue - Yes, I am aware of this problem. In fact, this is one problem, again, of `tf::_corun_until`. I am still researching a good way to solve this problem. The technical challenge we encounter is the data race:

Without using `tf::_corun_until`, the solution is to re-invoke `f1_work[m1]` when a worker finishes `mod_f3`:

1. w1 runs `f1_work[m1]` and proceed to the post work, and `m1` will start immediately
2. when the three tasks in `m1` all finish, `f1_work[m1]` will be re-invoked
3. w1 still wrapping up the post work `f1_work[m1]` while clashing the re-invoked `f1_work[m1]` - data race

Does this make sense to you? 

IMHO, the easies solution is the use of C++20 coroutine at which we can just save the context of w1 running `f1_work[m1]` so we don't have to worry about the data race with the post work. That being said, I am still thinking about how to solve this problem with a C++17 solution. In the meantime, I would appreciate if you can help contribute this example to our unittest :)

**Comment by ljwo** [comment]
It's good you know about the issue but it's a pity it is not documented (or else I missed it - please point me if I did miss it).

You say `enhancement`.

In process control this is not an enhancement. This is a disaster, a showstopper, a big red "no-go" sign. 

This leads to situations where you have parallel lines of execution with a task that performs a long operation on a synchronous api (like moving some parts of a machine or waiting on a timer) and some sequence of quick operations on another line. And now your sequence of quick operations is interrupted **for minutes** because the worker thread happened to steal the long running task and blocked the original line, so that no other worker thread can take it up, despite having enough processing resources (workers) in the system.

And it's preventable. Just don't use module tasks.

**Comment by tsung-wei-huang** [comment]
@ljwo can you try the dev branch? We have come up with a nice way to fix the problem. Module task is no longer blocking with corun.

**Comment by ljwo** [comment]
Works ok with the dev branch, your changes solve this issue and example 2 of https://github.com/taskflow/taskflow/issues/647.

**Comment by tsung-wei-huang** [comment]
@ljwo great! Thank you for pointing out this problem. I am doing some more tests and will merge it to the master branch and include it to the next release. In the meanwhile, please feel free to add your testcase to the [unit tests of module tasks](https://github.com/taskflow/taskflow/blob/dev/unittests/test_compositions.cpp)

**Comment by tsung-wei-huang** [comment]
Merged to the master branch. Will be included in the next release v3.9.
